packer/plugin/discover.go

```diff
@@ -0,0 +1,232 @@
+package plugin
+
+import (
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/packer/packer"
+	packersdk "github.com/hashicorp/packer/packer-plugin-sdk/packer"
+)
+
+// PACKERSPACE is used to represent the spaces that separate args for a command
+// without being confused with spaces in the path to the command itself.
+const PACKERSPACE = "-PACKERSPACE-"
+
+type Config struct {
+	PluginMinPort  int
+	PluginMaxPort  int
+	builders       packer.MapOfBuilder
+	provisioners   packer.MapOfProvisioner
+	postProcessors packer.MapOfPostProcessor
+}
+
+func (c *Config) GetPlugins() (packer.MapOfBuilder, packer.MapOfProvisioner, packer.MapOfPostProcessor) {
+	return c.builders, c.provisioners, c.postProcessors
+}
+
+// Discover discovers plugins.
+//
+// Search the directory of the executable, then the plugins directory, and
+// finally the CWD, in that order. Any conflicts will overwrite previously
+// found plugins, in that order.
+// Hence, the priority order is the reverse of the search order - i.e., the
+// CWD has the highest priority.
+func (c *Config) Discover() error {
+	c.builders = packer.MapOfBuilder{}
+	c.postProcessors = packer.MapOfPostProcessor{}
+	c.provisioners = packer.MapOfProvisioner{}
```

Comment:
I think we should init these only if they are nil ( I'm thinking of testing particularly where this _could_ cause issues ).

This could be fixed here, or at the first place these are used.

Reply:
Agree! I will do it in this same PR. Better to do it know 👍